src/chainstate/burn/operations/peg_out_fulfill.rs

```diff
@@ -0,0 +1,228 @@
+use stacks_common::codec::StacksMessageCodec;
+use stacks_common::types::chainstate::BlockHeaderHash;
+
+use crate::burnchains::BurnchainBlockHeader;
+use crate::burnchains::BurnchainTransaction;
+use crate::chainstate::burn::Opcodes;
+use crate::types::chainstate::StacksAddress;
+use crate::types::Address;
+
+use crate::chainstate::burn::operations::Error as OpError;
+use crate::chainstate::burn::operations::PegOutFulfillOp;
+
+impl PegOutFulfillOp {
+    pub fn from_tx(
+        block_header: &BurnchainBlockHeader,
+        tx: &BurnchainTransaction,
+    ) -> Result<Self, OpError> {
+        if tx.opcode() != Opcodes::PegOutFulfill as u8 {
+            warn!("Invalid tx: invalid opcode {}", tx.opcode());
+            return Err(OpError::InvalidInput);
+        }
+
+        let (amount, recipient) = if let Some(Some(recipient)) = tx.get_recipients().first() {
+            (recipient.amount, recipient.address.clone())
+        } else {
+            warn!("Invalid tx: Output 2 not provided");
+            return Err(OpError::InvalidInput);
+        };
+
+        let block_header_hash = Self::parse_data(&tx.data())?;
+
+        let txid = tx.txid();
+        let vtxindex = tx.vtxindex();
+        let block_height = block_header.block_height;
+        let burn_header_hash = block_header.block_hash;
+
+        Ok(Self {
+            block_header_hash,
+            amount,
+            recipient,
+            txid,
+            vtxindex,
+            block_height,
+            burn_header_hash,
+        })
+    }
+
+    fn parse_data(data: &[u8]) -> Result<BlockHeaderHash, ParseError> {
```

Comment:
This needs to be a `StacksBlockId`, not a `BlockHeaderHash`.

A Stacks block's bare hash alone does not refer to a unique chain tip.  Instead, it's the hash of the block, hashed with the `ConsensusHash` that represents the history of PoX operations leading to its block-commit.  This hash-of-hashes is captured within a `StacksBlockId`.

Reply:
Some elaboration.

In PoX, it is possible for the same Stacks block to be valid in two PoX forks.  This can happen if the same block is mined twice in two different PoX forks.  This can happen if their parent is in the prior reward cycle, but one child lives in a PoX fork where the anchor block was chosen and the other lives in a fork in which it was not chosen.

To address this corner case, we use `StacksBlockId` to represent the globally-unique ID for a block, since it represents the PoX history in which the block was mined.

Reply:
Ah I see. Thank you for the explanation. I will update the code to use `StacksBlockId`.

Reply:
Done :heavy_check_mark: